src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/primitives/impl/ControlPrimitives.java

```diff
@@ -856,15 +856,19 @@ protected final Object doNamedPrimitiveWithArgs(final VirtualFrame frame, @Suppr
     @GenerateNodeFactory
     @SqueakPrimitive(indices = 230)
     protected abstract static class PrimRelinquishProcessorNode extends AbstractPrimitiveNode implements BinaryPrimitive {
+        private static InterruptHandlerState istate;
```

Comment:
Static state is to be avoided. In the future, we want to support running multiple images on the same GraalVM.

Reply:
Well yes, but the current code is simply incorrect. It has guards that reference global mutable state (whether the external objects table is empty or not), and have nothing to do with the code object where they are used. As a result, asserts that the external objects is empty are generated, and the external objects table becomes non-empty dynamically at runtime, when an external semaphore is registered

Reply:
Could you please have a look at https://github.com/hpi-swa/graalsqueak/commit/902815fce0b4c4a961d054b6ddaf07caf5b3dd12? When I merged your recent change for `InterruptHandlerNode.java`, I didn't see the guard problem. It's well known and there are workarounds. Nonetheless, I think BranchProfiles are a better fit for this particular use case.

Reply:
Thanks. I don't know enough about Truffle to comment about that part.
I do want to mention that we can also avoid entering the external semaphores signalling loop by doing a quick check if InterruptHandlerState's semaphoresToSignal is empty